src/radical/pilot/agent/agent_0.py

```diff
@@ -404,63 +404,40 @@ def _start_sub_agents(self):
                                           'arguments'        : [bs_name, sa],
                                          }
                 }
-                cmd, hop = agent_lm.construct_command(agent_cmd,
-                        launch_script_hop='/usr/bin/env RP_SPAWNER_HOP=TRUE "%s"' % ls_name)
-
-                with open (ls_name, 'w') as ls:
-                    # note that 'exec' only makes sense if we don't add any
-                    # commands (such as post-processing) after it.
-                    ls.write('#!/bin/sh\n\n')
-                    for k,v in agent_cmd['description'].get('environment', {}).items():
-                        ls.write('export "%s"="%s"\n' % (k, v))
-                    ls.write('\n')
-                    for pe_cmd in agent_cmd['description'].get('pre_exec', []):
-                        ls.write('%s\n' % pe_cmd)
-                    ls.write('\n')
-                    ls.write('exec %s\n\n' % cmd)
-                    st = os.stat(ls_name)
-                    os.chmod(ls_name, st.st_mode | stat.S_IEXEC)
-
-                if hop : cmdline = hop
-                else   : cmdline = ls_name
-
-            # ------------------------------------------------------------------
-            class _SA(mp.Process):
-
-                def __init__(self, sa, cmd, log):
-                    self._name = sa
-                    self._cmd  = cmd.split()
-                    self._log  = log
-                    self._proc = None
-                    super(_SA, self).__init__(name=self._name)
-
-                    self.start()
-
-
-                def run(self):
-
-                    sys.stdout = open('%s.out' % self._name, 'w')
-                    sys.stderr = open('%s.err' % self._name, 'w')
-                    out        = open('%s.out' % self._name, 'w')
-                    err        = open('%s.err' % self._name, 'w')
-                    self._proc = sp.Popen(args=self._cmd, stdout=out, stderr=err)
-                    self._log.debug('sub-agent %s spawned [%s]', self._name,
-                            self._proc)
-
-                    assert(self._proc)
-
-                    # FIXME: lifetime, use daemon agent launcher
-                    while True:
-                        time.sleep(0.1)
-                        if self._proc.poll() is None:
-                            return True   # all is well
-                        else:
-                            return False  # proc is gone - terminate
-            # ------------------------------------------------------------------
+
+                # find a launcher to use
+                launcher = self._rm.find_launcher(agent_task)
+                if not launcher:
+                    raise RuntimeError('no launch method found for sub agent')
+
+                with open(launch_script, 'w') as fout:
+
+                    fout.write('#!/bin/sh\n\n')
```

Comment:
to have one `write` call (prepare text before to write it down)

p.s. for both launch- and exec-scripts

Reply:
Thanks, done.